src/clients/Deluge.ts

```diff
@@ -0,0 +1,232 @@
+import { InjectionResult } from "../constants.js";
+import { CrossSeedError } from "../errors.js";
+import { Label, logger } from "../logger.js";
+import { Metafile } from "../parseTorrent.js";
+import { getRuntimeConfig } from "../runtimeConfig.js";
+import { Searchee } from "../searchee.js";
+import { TorrentClient } from "./TorrentClient.js";
+import fetch, { Headers, Response } from "node-fetch";
+
+interface DelugeResponse {
+	error?: {
+		message?: string;
+		code?: number;
+	};
+	result?: any;
+}
+
+export default class Deluge implements TorrentClient {
+	private msgId = 0;
+	private delugeCookie = "";
+	private delugeWebUrl: URL;
+	private delugeLabel: string;
+	private isLabelEnabled: Promise<boolean>;
+
+	constructor() {
+		this.delugeWebUrl = new URL(`${getRuntimeConfig().delugeWebUrl}`);
+		this.delugeLabel = "cross-seed";
+	}
+
+	/**
+	 * validates the login and host for deluge webui
+	 */
+	async validateConfig(): Promise<void> {
+		const url = new URL(this.delugeWebUrl);
+		if (url.username && !url.password) {
+			throw new CrossSeedError(
+				"you need to define a password in the delugeWebUrl. (eg: http://:<PASSWORD>@localhost:8112)"
+			);
+		}
+		await this.authenticate();
+		this.isLabelEnabled = this.labelEnabled();
+	}
+
+	/**
+	 * connects and authenticates to the webui
+	 */
+	private async authenticate(): Promise<void> {
+		const response = await this.call({
+			method: "auth.login",
+			params: [this.delugeWebUrl.password],
+		});
+
+		if (response === null) {
+			throw new CrossSeedError(
+				`failed to establish a connection to deluge: ${this.delugeWebUrl.origin}`
+			);
+		}
+	}
+
+	/**
+	 * ensures authentication and sends JSON-RPC calls to deluge
+	 */
+	private async call(body: { method: string; params: any[] }) {
+		this.msgId = (this.msgId + 1) % 1024;
+
```

Comment:
I still like sendCall's old signature `sendCall(method: string, params: any[])` and I think we can still preserve that now that we only have `call`. 

```suggestion
	private async call(body: { method: string; params: any[] }) {
		this.msgId = (this.msgId + 1) % 1024;
		const requestBody: JSON.stringify({ method, params, id: this.msgId }),
```